pygmt/src/grdhisteq.py

```diff
@@ -0,0 +1,74 @@
+"""
+grdhisteq - Compute directional gradients from a grid.
+"""
+
+import xarray as xr
+from pygmt.clib import Session
+from pygmt.helpers import (
+    GMTTempFile,
+    build_arg_string,
+    fmt_docstring,
+    kwargs_to_strings,
+    use_alias,
+)
+
+
+@fmt_docstring
+@use_alias(
+    C="divisions",
+    G="outgrid",
+    R="region",
+    N="gaussian",
+    Q="quadratic",
+    V="verbose",
+    n="interpolation",
+)
+@kwargs_to_strings(R="sequence")
+def grdhisteq(grid, **kwargs):
+    r"""
+    Perform histogram equalization for a grid.
+
+    Full option list at :gmt-docs:`grdhisteq.html`
+
+    {aliases}
+
+    Parameters
+    ----------
+    grid : str or xarray.DataArray
+        The file name of the input grid or the grid loaded as a DataArray.
+    outgrid : str or None
+        The name of the output netCDF file with extension .nc to store the grid
+        in.
+    divisions : int or str
+        The number of divisions of data range to make [Default is 16].
+
+    {R}
+    {V}
+
+    Returns
+    -------
+    ret: xarray.DataArray or None
+        Return type depends on whether the ``outgrid`` parameter is set:
+
+        - :class:`xarray.DataArray` if ``outgrid`` is not set
+        - None if ``outgrid`` is set (grid output will be stored in file set by
+          ``outgrid``)
+    """
+    with GMTTempFile(suffix=".nc") as tmpfile:
+        with Session() as lib:
+            file_context = lib.virtualfile_from_data(check_kind="raster", data=grid)
+            with file_context as infile:
+                if "G" not in kwargs.keys():  # if outgrid is unset, output to tempfile
+                    kwargs.update({"G": tmpfile.name})
+                outgrid = kwargs["G"]
+                arg_str = " ".join([infile, build_arg_string(kwargs)])
+                lib.call_module("grdhisteq", arg_str)
+
+        if outgrid == tmpfile.name:  # if user did not set outgrid, return DataArray
+            with xr.open_dataarray(outgrid) as dataarray:
+                result = dataarray.load()
+                _ = result.gmt  # load GMTDataArray accessor information
+        else:
+            result = None  # if user sets an outgrid, return None
+
+        return result
```

Comment:
I added support for table output based on your triangulate implementation. I still need to add a test.

Reply:
Regarding this and your comment that `-h` is supported, we probably have a bug with the functions that include `result = pd.read_csv(tmpfile.name, sep="\t", header=None)` and `h="header"` since the user could specify that GMT should add a header to the output.

Reply:
Right, so we might need to do something like `header=None if "h" not in kwargs else kwargs["h"]` (or something more complicated). Also, I just noticed `grdhisteq -h` is not in GMT 6.2, only GMT 6.3, so should we include it here or not?

Reply:
If you type `gmt grdhisteq -?` in GMT 6.2, you will see that `-h` was supported. I have been going through making sure that the UNIX manpage-type docs are consistent with the html documentation, so there may be some options that look new based on the online reference but are just doc fixes.

Reply:
> Right, so we might need to do something like `header=None if "h" not in kwargs else kwargs["h"]` (or something more complicated)

We might need something more complicated since GMT can write multi-row headers depending on the modifiers. I think I'll just leave ``header`` off this PR and deal with it separately.

Reply:
Yep, the `-h` option will need to handled properly for other table functions too, so best to do that separately.